aten/src/ATen/native/cuda/UpSampleNearest2d.cu

```diff
@@ -133,70 +204,113 @@ static void upsample_nearest2d_out_cuda_template(
 
-  if (input.numel() == 0) {
+  if (input_.sizes() == output.sizes()) {
+    output.copy_(input_);
     return;
   }
 
-  int nc = nbatch * channels;
-
-  const int max_threads = std::min<int>(
-      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, MAX_THREADS);
-
-  int* maxThreadsDim = at::cuda::getCurrentDeviceProperties()->maxThreadsDim;
-  int* maxGridSize = at::cuda::getCurrentDeviceProperties()->maxGridSize;
-
-  // upsample_nearest2d meta call makes sure input/output tensor is not empty;
-  int block_x = std::min<int>(
-      maxThreadsDim[0], std::min<int>(lastPow2(output_width), max_threads));
-  int block_y = std::min<int>(
-      maxThreadsDim[1],
-      std::min<int>(lastPow2(output_height), max_threads / block_x));
-  int block_z = std::min<int>(
-      maxThreadsDim[2], std::min<int>(nc, max_threads / block_x / block_y));
-  const dim3 block(block_x, block_y, block_z);
-
-  int grid_x = cuda::ATenCeilDiv(output_width, block_x);
-  int grid_y = cuda::ATenCeilDiv(output_height, block_y);
-  int grid_z = std::min<int>(
-      maxGridSize[2], cuda::ATenCeilDiv(nc, block_z * 4));
-  const dim3 grid(grid_x, grid_y, grid_z);
-  // Error out on cases where grid_x & grid_y exceeds limit of launch config, as
-  // the current kernel implementation doesn't loop over the two dimensions.
-  // This is unlikely to happen.
-  // TODO: kernel implementation could stride on spatial dimension. We probably
-  //       need to overhaul the kernel.
-  TORCH_CHECK(
-      grid_x <= maxGridSize[0] && grid_y <= maxGridSize[1],
-      "input tensor has spatial dimension larger than the kernel capacity");
-
-  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-  AT_DISPATCH_FLOATING_TYPES_AND2(ScalarType::Half, ScalarType::Byte, input.scalar_type(), "upsample_nearest2d_out_frame", [&] {
-        using accscalar_t = at::acc_type<scalar_t, true>;
-
-        auto idata = input.data_ptr<scalar_t>();
-        auto odata = output_c.data_ptr<scalar_t>();
-
-        const float height_scale = compute_scales_value<float>(scales_h, input_height, output_height);
-        const float width_scale = compute_scales_value<float>(scales_w, input_width, output_width);
-
-        upsample_nearest2d_out_frame<scalar_t, accscalar_t>
-            <<<grid, block, 0, stream>>>(
-                idata,
-                odata,
-                nc,
-                input_height,
-                input_width,
-                output_height,
-                output_width,
-                height_scale,
-                width_scale);
-        C10_CUDA_KERNEL_LAUNCH_CHECK();
-      });
-
-  if (!output.is_contiguous()) {
-      output.copy_(output_c);
+  // heuristic: only use channels_last path when it's faster than the contiguous path
+  if (memory_format == at::MemoryFormat::ChannelsLast && channels >= 4) {
+  // if (false) {
+    TORCH_INTERNAL_ASSERT(output.is_contiguous(at::MemoryFormat::ChannelsLast),
```

Comment:
`upsample_nearest2d_out_cuda` is a user facing function, which means that someone can call it with discontiguous `out`. So here, instead of TORCH_INTERNAL_ASSERT, you should have handling of discontiguous `out` (by creating contiguous intermediate, and copying to provided `out`). Same for bilinear, I think

Reply:
fixed